File: python/nodeSwap.py
class LinkedList:
    def __init__(self, value):
        self.value = value
        self.next = None
        

# Time -> 0(N) Space -> 0(1)
def nodeSwap(head):
    swap = head
    
    while swap != None:
        if swap.next == None:
            break
        swap.value, swap.next.value = swap.next.value, swap.value
        swap = swap.next.next 
    
    return head


head = LinkedList(0)
curr = head
array = [3, 0, 5, 2, 1]

for val in range(len(array)):
    curr.value = array[val]
    if val == len(array) - 1:
        break
    curr.next = LinkedList(0)
    curr = curr.next
    
arrange = nodeSwap(head)

while arrange != None:
    print(arrange.value, end=" -> ")
    arrange = arrange.next
print()

File: go/nodeswap/nodeSwap.go
package main

import "fmt"

type LinkedList struct {
	Value int
	Next *LinkedList
}


// Time -> 0(N) Space -> 0(1)
func NodeSwap(head *LinkedList) *LinkedList {
	loop := head

	for loop != nil {
		if loop.Next == nil {break}
		loop.Value, loop.Next.Value = loop.Next.Value, loop.Value
		loop = loop.Next.Next
	}

	return head
}

func main() {

	list := []int{7, 8, 9, 10, 0}
	head := &LinkedList{}
	curr := head
	for i, val := range list {
		curr.Value = val
		if i == len(list) - 1 {continue}
		curr.Next = &LinkedList{}
		curr = curr.Next
	}

	rearranged := NodeSwap(head)
	for rearranged != nil {
		fmt.Printf("%v -> ", rearranged.Value)
		rearranged = rearranged.Next
	}
	fmt.Println()
}
